livekit-plugins/livekit-plugins-assemblyai/livekit/plugins/assemblyai/stt.py

```diff
@@ -0,0 +1,395 @@
+                    # and should be replaced with a continuous resampling
+                    frame = data.remix_and_resample(
+                        self._opts.sample_rate,
+                        self._num_channels,
+                    )
+                    buffer.extend(frame.data.tobytes())
+
+                    # Calculate buffer duration
+                    total_frames = len(buffer) / bytes_per_frame[self._opts.encoding]
+                    samples_per_second = self._opts.sample_rate * self._num_channels
+                    buffer_duration = total_frames / samples_per_second
+
+                    if buffer_duration >= self._opts.buffer_size_seconds:
+                        await ws.send_bytes(bytes(buffer))
+                        buffer.clear()
+                        buffer_duration = 0.0
+                elif data == END_UTTERANCE_SILENCE_THRESHOLD_MSG:
+                    await ws.send_str(data)
+                elif data == SpeechStream._CLOSE_MSG:
+                    closing_ws = True
+                    await ws.send_str(data)  # tell AssemblyAI we are done with inputs
+                    break
+                else:
+                    raise ValueError("Received unexpected data type: ", type(data))
+
+        async def recv_task():
+            nonlocal closing_ws
+            while True:
+                msg = await ws.receive()
+                if msg.type in (
+                    aiohttp.WSMsgType.CLOSED,
+                    aiohttp.WSMsgType.CLOSE,
+                    aiohttp.WSMsgType.CLOSING,
+                ):
+                    if closing_ws:  # close is expected, see SpeechStream.aclose
+                        return
+
+                    raise Exception(
+                        "AssemblyAI connection closed unexpectedly",
+                    )  # this will trigger a reconnection, see the _run loop
+
+                if msg.type != aiohttp.WSMsgType.TEXT:
+                    logger.error("unexpected AssemblyAI message type %s", msg.type)
+                    continue
+
+                try:
+                    # received a message from AssemblyAI
+                    data = json.loads(msg.data)
+                    self._process_stream_event(data)
+                except Exception:
+                    logger.error("failed to process AssemblyAI message")
+
+        await asyncio.gather(send_task(), recv_task())
+
+    def _end_speech(self) -> None:
+        if not self._speaking:
+            logger.info(
+                "trying to commit final events without being in the speaking state",
+            )
+            return
+
+        if len(self._final_events) == 0:
+            return
+
+        self._speaking = False
+
+        # combine all final transcripts since the start of the speech
+        sentence = ""
+        confidence = 0.0
+        for f in self._final_events:
+            alt = f.alternatives[0]
+            sentence += f"{alt.text.strip()} "
```

Comment:
nit pick but I think this can be simplified to
`sentence = " ".join(f.alternatives[0].text.strip() for f in self._final_events)`
and then you don't need the rstrip below.
Even if you need a second loop for the confidence, I think this solution is more efficient. You could even use the `sum` function to have another one-liner for the confidence